product/crud/handlers/handle_create_product.py

```diff
@@ -1,51 +1,46 @@
-from product.crud.handlers.utils.http_responses import build_response
 from product.crud.handlers.utils.observability import logger, metrics, tracer
-from product.crud.schemas.exceptions import InternalServerException, ProductAlreadyExistsException
-from product.crud.schemas.input import CreateProductRequest
+from product.crud.handlers.utils.rest_api_resolver import app
+from product.crud.schemas.input import CreateProductInput
 from product.crud.schemas.output import CreateProductOutput
 
 
-@init_environment_variables(model=CreateVars)
-@metrics.log_metrics
-@tracer.capture_lambda_handler(capture_response=False)
-def handle_create_product(event: Dict[str, Any], context: LambdaContext) -> Dict[str, Any]:
-    logger.set_correlation_id(context.aws_request_id)
-
+@app.route(PRODUCT_PATH, method=HTTPMethod.PUT)
+def handle_create_product(product_id: str) -> dict[str, Any]:
     env_vars: CreateVars = get_environment_variables(model=CreateVars)
     logger.debug('environment variables', extra=env_vars.model_dump())
 
-    try:
-        # we want to extract and parse the HTTP body from the api gw envelope
-        create_input: CreateProductRequest = parse(event=event, model=CreateProductRequest)
-        logger.info('got create product request', extra={'product': create_input.model_dump()})
-    except (ValidationError, TypeError) as exc:  # pragma: no cover
-        logger.exception('event failed input validation', extra={'error': str(exc)})
-        return build_response(http_status=HTTPStatus.BAD_REQUEST, body={})
+    # we want to extract and parse the HTTP body from the api gw envelope
+    payload = app.current_event.body if app.current_event.body is not None else ''
+    create_input: CreateProductInput = CreateProductInput.model_validate_json(payload)
 
+    logger.info('got a valid create product request', extra={'product': create_input.model_dump(), 'product_id': product_id})
     metrics.add_metric(name='CreateProductEvents', unit=MetricUnit.Count, value=1)
-    try:
-        response: CreateProductOutput = create_product(
-            product_id=create_input.pathParameters.product,
-            product_name=create_input.body.name,
-            product_price=create_input.body.price,
-            table_name=env_vars.TABLE_NAME,
-        )
-    except InternalServerException:
-        logger.exception('finished handling create product request with internal error')
-        return build_response(http_status=HTTPStatus.INTERNAL_SERVER_ERROR, body={})
-    except ProductAlreadyExistsException:  # pragma: no cover
-        logger.exception('finished handling create product request with bad request')
-        return build_response(http_status=HTTPStatus.BAD_REQUEST, body={'error': 'product already exists'})
-
-    logger.info('finished handling create product request, product created')
-    return build_response(http_status=HTTPStatus.OK, body=response.model_dump())
+
+    response: CreateProductOutput = create_product(
```

Comment:
same - is mypy unable to infer?

That's for my own learning purposes with Pydantic V2 and mypy pydantic plugin

Reply:
so, i just like to add it to make it clearer here

Reply:
gotcha!